### PR TITLE
Fix and improve list parser of cabal init cli

### LINKED ISSUE #8659: Cabal init doesn't honour multiple `--dependency` flags
**Describe the bug**
`cabal init --lib -d base -d vector` should create a cabal file containing both of these dependencies.

**To Reproduce**

```
$ cabal init --lib -d base -d vector
$ cat test.cabal
...
    build-depends:    vector
```
but I'd expect:
```
$ cabal init --lib -d base -d vector
$ cat test.cabal
...
    build-depends:    base, vector
```

**System information**
 - `cabal 3.6`, `cabal 3.8` and cabal HEAD

### PR DESCRIPTION
**Pull request by fendor**
Fixes #8659 

Fix cli list parser errors in `cabal init`

Occurrences of `Flag [a]` seem to behave in an unexpected way. The monoid
instance of `Flag` is right associative and discard the value on the
left. 

Fixes for all `Flag [a]` the cli parser in `cabal init`. Adds tests for proving the improvement.

### CODE REVIEW

File: cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs [hunk 1]
```diff
@@ -1265,3 +1271,35 @@ testGo label f g h inputs = testCase label $
...
+        ["-d", "base,vector"]
+      flags @?= dummyFlags
+        { dependencies = fmap mkStringyDep ["base", "vector"]
+        }
+  , testCase "Fail to parse multiple exposed-modules" $ do
+      flags <- assumeAllParse $ commandParseArgs initCommand False
+        ["-o", "Test", "-o", "Test2"]
+      flags @?= dummyFlags
+        { exposedModules = Flag (fmap fromString ["Test2"])
+        }
```
Comment (fendor) [inline cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs:1299]: This test-case proves the same problem for all other occurrences of `Flag [a]` in `InitFlags`.

File: cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs [hunk 2]
```diff
@@ -1265,3 +1271,35 @@ testGo label f g h inputs = testCase label $
...
+  , testCase "Parse multiple dependency flags" $ do
+      flags <- assumeAllParse $ commandParseArgs initCommand False
+        ["-d", "base", "-d", "vector"]
+      flags @?= dummyFlags
+        { dependencies = fmap mkStringyDep ["base", "vector"]
+        }
+  , testCase "Parse list of dependencies" $ do
+      flags <- assumeAllParse $ commandParseArgs initCommand False
+        ["-d", "base,vector"]
```
Comment (fendor) [inline cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs:1290]: Sneakily introduced new feature.

Comment (Mikolaj) [reply cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs:1290]: I'd rename the PR to "Fix and improve" and keep it.

Comment (Mikolaj) [reply cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs:1290]: and document in many places...

Comment (fendor) [reply cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs:1290]: Any idea where this might be documented?

**Comment by fendor** [comment]
@emilypi Tests fail because tests use the state `Flag []` to express something that I don't consider possible in the interactive and non-interactive execution. Namely, it uses `Flag []` to avoid an entry in `_libDependencies`, allowing the following to hold: `_libDependencies @?= []`. However, when we use `[]`, we now add the `base` dependency, forcing the assertion to become `_libDependencies @?= [Dependency (PackageName "base") (MajorBoundVersion (mkVersion [4,16,4,0])) (fromNonEmpty (LMainLibName :| []))]`.

I think the new assertion is the more correct form since it is actually what's going to be generated. Any opinions?

**Comment by fendor** [comment]
Ready for review, just fix the parsers, and add test-cases for all modified parsers.

File: cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs (hunk 2 above)
Comment (fendor) [reply cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs]: No idea where to document that, the flag doesn't seem to be documented in the .rst.

**Review by emilypi — APPROVED** [review]
LGTM thanks @fendor

File: cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs (hunk 2 above)
Comment (Mikolaj) [reply cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs]: Well, anywhere. In a changelog for starters. Perhaps in the `--help` output? Would it be awkward a sentence into the .rst?

**Review by Mikolaj — CHANGES_REQUESTED** [review]
Thanks a lot. I'm marking the absence of a changelog so that we don't miss it when reviewing again.

File: cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs (hunk 2 above)
Comment (fendor) [reply cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs]: in the .rst, the options aren't explained at all. Thus, adding docs for only a subset of flags feels like a bad decision to me (at least I resist documenting the whole command right now :) ). Changelog and help message make sense, thank you very much!

Comment (fendor) [reply cabal-install/tests/UnitTests/Distribution/Client/Init/NonInteractive.hs]: Done!

**Review by Mikolaj — APPROVED** [review]
Brilliant.

**Comment by Mikolaj** [comment]
@fendor: when you are ready, feel free to set a label (I guess squash+merge_me?).

**Comment by Mikolaj** [comment]
This is the only PR left before cutting 3.10, so let me speed it up...

**Pull request merged**